google/cloud/bigquery/_helpers.py

```diff
@@ -1004,3 +1005,61 @@ def _verify_job_config_type(job_config, expected_type, param_name="job_config"):
                 job_config=job_config,
             )
         )
+
+
+class ResourceBase:
+    """Base class providing the from_api_repr method."""
+
+    def __init__(self):
+        self._properties = {}
+
+    @classmethod
+    def from_api_repr(cls, resource: dict):
+        """Factory: constructs an instance of the class (cls)
+        given its API representation.
+
+        Args:
+            resource (Dict[str, Any]):
+                API representation of the object to be instantiated.
+
+        Returns:
+            ResourceBase: An instance of the class initialized with data
+                from 'resource'.
+        """
+        config = cls()
+        config._properties = copy.deepcopy(resource)
+        return config
+
+
+def _isinstance_or_raise(value: Any, dtype: Optional[Union[Any, Tuple[Any, ...]]]):
+    """Determine whether a value type matches a given datatype or None.
+
+    Args:
+        value (Any): Value to be checked.
+        dtype (Optional[Union[Any, Tuple[Any, ...]]]): Expected data type(s).
+
+    Returns:
+        Any: Returns the input value if the type check is successful.
+
+    Raises:
+        TypeError: If the input value's type does not match the expected data type(s).
+    """
+    
+    # Simplest case
+    if dtype is None and value is None:
+        return value
+    
+    elif isinstance(dtype, tuple):
```

Comment:
I think the built-in `isinstance()` supports checking multiple types: https://stackoverflow.com/questions/33311258/python-check-if-variable-isinstance-of-any-type-in-list

Reply:
You are absolutely right, `isinstance()` will check multiple types if you supply a `tuple`.

But it does not correctly handle the object `None`. 

*That is one of the main reasons this function has so many checks, because we are adding to the capability of the built-in `isinstance()` functionality.*

```
In [5]: isinstance(dtype, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 isinstance(dtype, None)

TypeError: isinstance() arg 2 must be a type, a tuple of types, or a union
```

Interestingly, `isinstance()` **shortcircuits** (stops all checks as soon as one passes), so it is very easy to accidentally believe that it is behaving as one expects.

Here it does not balk unless `None` is first, cause it succeeds with `int` in the other cases and never checks against `None`.

```
In [8]: test_value = 42

In [9]: isinstance(test_value, (int, float, None))
Out[9]: True

In [11]: isinstance(test_value, (int, None, float))
Out[11]: True
```

But here it freaks out:

```
In [10]: isinstance(test_value, (None, int, float))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[10], line 1
----> 1 isinstance(test_value, (None, int, float))
```

Reply:
This is also why my test cases include tuples and in the tuples include None in the first and last position to ensure that our function behaves as expected in all cases.

```
(None, None, None),
('hello world.uri', str, 'hello world.uri'),
(None, (str, None), None),
(None, (None, str), None),
(None, (str, None), None),
('hello world.uri', (None, str), 'hello world.uri'),
('hello world.uri', (str, None), 'hello world.uri'),
```